src/model.rs

```diff
@@ -47,45 +49,115 @@ impl Marginal {
                     && fraction > NotNan::new(0.0).unwrap()
                     && fractions.len() > 1
                 {
-                    // only if the fraction for the current has greater than 0.0
-                    let mut haplotype_group = Haplotype("".to_string());
-                    let current_haplotype = &self.haplotypes[haplotype_index];
                     if self.application == "hla".to_string() {
+                        // only if the fraction for the current has greater than 0.0
+                        let current_haplotype = &self.haplotypes[haplotype_index];
                         let splitted = &self.haplotypes[haplotype_index]
                             .split(':')
                             .collect::<Vec<&str>>();
-                        haplotype_group = Haplotype(splitted[0].to_owned() + &":" + splitted[1]);
+                        let mut haplotype_group =
+                            Haplotype(splitted[0].to_owned() + &":" + splitted[1]);
+
+                        //find the index of the (haplotype, haplotype_group) in graph
+                        if let Some(haplotype_graph) = &self.haplotype_graph {
+                            let index = haplotype_graph
+                                .node_indices()
+                                .find(|i| {
+                                    &haplotype_graph[*i]
+                                        == &(current_haplotype.clone(), haplotype_group.clone())
+                                })
+                                .unwrap();
```

Comment:
Doing something with a potentially linear runtime inside of the recursion will make things unnecessarily slow. Instead you should generate a lookup table to get the node index before (e.g. when initializing the marginal or rather directly inside of the haplotype_graph object, with a corresponding method that you can just call here to do the lookup in the table.

Reply:
This is improved now.